Performance/QPSBenchmark/Sources/QPSBenchmark/Runtime/Async/AsyncQPSServerImpl.swift

```diff
@@ -0,0 +1,99 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import Foundation
+import GRPC
+import Logging
+import NIOCore
+import NIOPosix
+
+/// Server setup for asynchronous requests.
+final class AsyncQPSServerImpl: AsyncQPSServer {
+  private let eventLoopGroup: MultiThreadedEventLoopGroup
+  private let server: Server
+  private let threadCount: Int
+
+  private var statsPeriodStart: DispatchTime
+  private var cpuStatsPeriodStart: CPUTime
+
+  private let logger = Logger(label: "AsyncQPSServerImpl")
+
+  /// Initialisation.
+  /// - parameters:
+  ///     - config: Description of the type of server required.
+  init(config: Grpc_Testing_ServerConfig) async throws {
+    // Setup threads as requested.
+    let threadCount = config.asyncServerThreads > 0
+      ? Int(config.asyncServerThreads)
+      : System.coreCount
+    self.threadCount = threadCount
+    self.logger.info("Sizing AsyncQPSServerImpl", metadata: ["threads": "\(threadCount)"])
+    self.eventLoopGroup = MultiThreadedEventLoopGroup(numberOfThreads: threadCount)
+
+    // Start stats gathering.
+    self.statsPeriodStart = grpcTimeNow()
+    self.cpuStatsPeriodStart = getResourceUsage()
+
+    let workerService = AsyncBenchmarkServiceImpl()
+
+    // Start the server
+    self.server = try await Server.insecure(group: self.eventLoopGroup)
+      .withServiceProviders([workerService])
+      .withLogger(self.logger)
+      .bind(host: "localhost", port: Int(config.port))
+      .get()
+  }
+  
+  func getServerInfo() -> ServerInfo {
+    let port = self.server.channel.localAddress?.port ?? 0
+    return ServerInfo(threadCount: self.threadCount, port: port)
+  }
+
+  /// Send the status of the current test
+  /// - parameters:
+  ///     - reset: Indicates if the stats collection should be reset after publication or not.
+  ///     - responseStream: response stream to which the status should sent to.
+  func sendStatus(reset: Bool, responseStream: GRPCAsyncResponseStreamWriter<Grpc_Testing_ServerStatus>) async throws {
+      let currentTime = grpcTimeNow()
+      let currentResourceUsage = getResourceUsage()
+      var result = Grpc_Testing_ServerStatus()
+      result.stats.timeElapsed = (currentTime - self.statsPeriodStart).asSeconds()
+      result.stats.timeSystem = currentResourceUsage.systemTime - self.cpuStatsPeriodStart
+        .systemTime
+      result.stats.timeUser = currentResourceUsage.userTime - self.cpuStatsPeriodStart.userTime
+      result.stats.totalCpuTime = 0
+      result.stats.idleCpuTime = 0
+      result.stats.cqPollCount = 0
+      self.logger.info("Sending server status")
+      try await responseStream.send(result)
+      if reset {
+        self.statsPeriodStart = currentTime
+        self.cpuStatsPeriodStart = currentResourceUsage
+      }
+  }
+
+  /// Shut down the service.
+  func shutdown() async throws {
+    do {
+      try await self.server.close().get()
```

Comment:
There should be an `initiateGracefulShutdown()` API (or something along those lines), we should use that instead as it allows active RPCs to finish etc.